libs/algorithms/include/hpx/parallel/algorithms/for_each.hpp

```diff
@@ -434,117 +608,175 @@ namespace hpx { namespace parallel { inline namespace v1 {
+                hpx::parallel::traits::is_indirect_callable<hpx::parallel::execution::sequenced_policy, F,
+                    hpx::parallel::traits::projected<Proj, InIter>>::value)
+#endif
+            >
+        // clang-format on
+        friend F tag_invoke(hpx::for_each_t, InIter first, InIter last, F&& f,
+            Proj&& proj = Proj())
+        {
+            using is_segmented = hpx::traits::is_segmented_iterator<InIter>;
+
+            parallel::v1::detail::for_each_(hpx::parallel::execution::seq,
+                first, last, f, std::forward<Proj>(proj), is_segmented());
+            return std::forward<F>(f);
+        }
+
+        // clang-format off
+        template <typename ExPolicy, typename FwdIter,
+            typename F, typename Proj = parallel::util::projection_identity,
+            HPX_CONCEPT_REQUIRES_(
+                hpx::parallel::execution::is_execution_policy<ExPolicy>::value &&
+                hpx::traits::is_forward_iterator<FwdIter>::value &&
+                hpx::parallel::traits::is_projected<Proj, FwdIter>::value)
+#if (!defined(__NVCC__) && !defined(__CUDACC__)) || defined(__CUDA_ARCH__)
+                ,
+            HPX_CONCEPT_REQUIRES_(
+                hpx::parallel::traits::is_indirect_callable<ExPolicy, F,
+                    hpx::parallel::traits::projected<Proj, FwdIter>>::value)
+#endif
+            >
+        // clang-format on
+        friend typename parallel::util::detail::algorithm_result<ExPolicy,
+            void>::type
+        tag_invoke(hpx::for_each_t, ExPolicy&& policy, FwdIter first,
+            FwdIter last, F&& f, Proj&& proj = Proj())
+        {
+            using is_segmented = hpx::traits::is_segmented_iterator<FwdIter>;
+
+            return parallel::v1::detail::get_for_each_void_result<>(
+                parallel::v1::detail::for_each_(std::forward<ExPolicy>(policy),
+                    first, last, std::forward<F>(f), std::forward<Proj>(proj),
+                    is_segmented()));
+        }
+    } for_each{};
+
+    HPX_INLINE_CONSTEXPR_VARIABLE struct for_each_n_t final
+      : hpx::functional::tag<for_each_n_t>
+    {
+    private:
+        // clang-format off
+        template <typename InIter, typename Size, typename F,
+            typename Proj = parallel::util::projection_identity,
+            HPX_CONCEPT_REQUIRES_(
+                hpx::traits::is_input_iterator<InIter>::value &&
+                hpx::parallel::traits::is_projected<Proj, InIter>::value &&
+                hpx::parallel::traits::is_indirect_callable<hpx::parallel::execution::sequenced_policy, F,
+                    hpx::parallel::traits::projected<Proj, InIter>>::value
+            )>
+        // clang-format on
+        friend typename parallel::util::detail::algorithm_result<
+            hpx::parallel::execution::sequenced_policy, InIter>::type
+        tag_invoke(hpx::for_each_n_t, InIter first, Size count, F&& f,
+            Proj&& proj = Proj())
```

Comment:
The non-policy for_each/for_each_n can simply forward to `std::for_each` and `std::for_each_n`.

Reply:
I've left these to use `hpx::parallel::detail::for_each_` as `std::for_each` is C++17.